kubernetes/config/kube_config.c

```diff
@@ -175,15 +204,75 @@ static const kubeconfig_property_t *kubeconfig_get_current_property(kubeconfig_p
     return current_property;
 }
 
+static int kubeconfig_exec(kubeconfig_property_t * current_user)
+{
+    static char fname[] = "kubeconfig_exec()";
+    int rc = 0;
+
+    ExecCredential_t *exec_output = exec_credential_create();
+    if (!exec_output) {
+        fprintf(stderr, "%s: Cannot allocate memory for exec credential.[%s]\n", fname, strerror(errno));
+        return -1;
+    }
+
+    rc = kube_exec_and_get_result(exec_output, current_user->exec);
+    if (0 != rc) {
+        fprintf(stderr, "%s: The kubeconfig exec failed.\n", fname);
+        goto end;
+    }
+    if (!exec_output->status) {
+        rc = -1;
+        fprintf(stderr, "%s: The kubeconfig exec result is not valid.\n", fname);
+        goto end;
+    }
+    if (EXEC_CREDENTIAL_TYPE_TOKEN == exec_output->status->type && exec_output->status->token) {
+        current_user->token = strdup(exec_output->status->token);
+    } else if (exec_output->status->clientCertificateData && exec_output->status->clientKeyData) {
+        current_user->client_certificate_data = strdup(exec_output->status->clientCertificateData);
+        current_user->client_key_data = strdup(exec_output->status->clientKeyData);
+    } else {
+        rc = -1;
+        fprintf(stderr, "%s: Cannot get the authentication infomation from the kubeconfig exec result.\n", fname);
+        goto end;
+    }
+
+  end:
+    exec_credential_free(exec_output);
+    exec_output = NULL;
+    return rc;
+}
+
+static int kubeconfig_update_exec_command_path(kubeconfig_property_t * exec, const char *kube_config_file)
+{
+    static char fname[] = "kubeconfig_update_exec_command_path()";
+
+    if (!exec->command || 0 == strlen(exec->command)) {
+        return 0;
+    }
+
+    if ('/' != exec->command[0]) {  // relative path e.g. "./bin/" or "bin/"
+        const char *kube_config_dirname = dirname(kube_config_file);
+        char *original_command = exec->command;
+        int new_command_length = strlen(kube_config_dirname) + strlen("/") + strlen(original_command) + 1 /* 1 for the terminal of string */ ;
+        exec->command = calloc(1, new_command_length);
+        if (!exec->command) {
+            fprintf(stderr, "%s: Cannot allocate memory for exec command.[%s]\n", fname, strerror(errno));
```

Comment:
this is going to leak the memory in `original_command`

Reply:
OK. I will handle it.